### ISSUE TITLE
Add async copy buffering

### ISSUE THREAD

**Issue by klauspost** [ncw/rclone#164]
When I am uploading files (HD->ACD), I find that when there is some additional activity on the drive will lower my upload rate considerably.

I have found the root issue to be that the read size by each output is passed through unchanged to the underlying input. This means that if the output is copying small sizes that is also reflected on the input.

## Solution A
A one-line fix for this would be to slap on a bufio.Reader with a buffer size of, say a megabyte.

## Solution B
A better, but slightly more complex solution would be to have an asynchronous buffer, where the input reader is running all the time, unless the buffer is full. This would mean that output is not blocked by input and vice versa.

This would mean that for all transfers that if either input or output is bursty, the overall transfer will not be affected as much. 

The buffer size could be something like 4x4MB per transfer.

**Comment by ncw** [comment]
I spent some time thinking about this sort of thing earlier this year - there was [a long thread on go-nuts](https://groups.google.com/d/msg/golang-nuts/Mwn9buVnLmY/HFMRQm8JPNkJ) about it.  The conclusion was that the async buffer is really hard to get right under all circumstances.  It didn't cross my mind that this is relevant to `rclone` but it certainly is!

Here is a repo of code which came out of that discussion

  * https://godoc.org/github.com/karalabe/bufioprop

Probably the highest performing solution would be to use a buffered pipe and an io.Copy in a go routine to decouple the reading from the writing and batch them up into bigger batches.

**Comment by klauspost** [comment]
Ah yes, now I remember that. Unless I am mistaking, if the input is read-ahead buffered we should have enough "buffering" to not need the copy.

I've asked on the [Go forum](https://forum.golangbridge.org/t/async-buffered-reader-package/1126/4) if anyone has a package for this, but i might try to cook one up.

**Comment by ncw** [comment]
The bufioprop package has the right logic in, but not quite in the right form - the implementation is worth a look. It might be reading from net or disk and writing to net or disk so getting it to run quickly in all cases might be tricky.

**Comment by ncw** [comment]
I've added some general comments on the commit 2b83830401cb1516f2fd3744623907b9c120d9d3

I think the code looks very nice.

There are 3 magic numbers it would be nice to have a better justification for 
  * the cutoff size before we engage the buffer
  * the size of buffer
  * the number of buffers

A bit of testing with rclone could help.  Probably want 10s of largish file so rclone can transfer them in parallel and measure the throughput for
  * disk -> disk
  * disk -> remote
  * remote -> disk
  * remote -> remote

I wonder whether it might be better to have a larger number of smaller (say 32k) buffers. These could then be managed with a `sync.Pool` which would mean that they wouldn't be used if necessary (eg when the output is keeping up with the input). The buffer size should certainly be a multiple of the page size.

Another thing I was wondering was whether we should implement the [ReaderAt](https://golang.org/pkg/io/#ReaderAt) interface which would save copying the data?

**Comment by klauspost** [comment]
I did a bit of provisional benchmarking, and getting some hard numbers beyond what I have presented is tricky. The buffers are there mainly to handle uneven and smooth out bursty reads/writes. I cannot think of a good way to benchmark that. I think that no matter how we benchmark we will just end up optimizing for the specific benchmark.

So for me, the goal is to make sure there aren't any regression cases, while improving a proven worst case.

The gist of this is dealing with spinning disk thrashing and OS read scheduling. If we just test what you propose, there isn't going to be any significant changes unless you start pressuring the OS, by reading something else while you are running rclone.

I verified this by running 10GB HD->SSD, HD->HD and SSD->HD rclone copy with and without the change. There wasn't any significant difference. There are of course rather high variance in these tests, since harddisk thrashing isn't predictable. So local file-copy at least has no regressions.

I also tried benchmarking HD->SSD rclone while running a copy from the HD. It is depending on a *lot* of factors.

## Master when copying files:
![master-with-copy](https://cloud.githubusercontent.com/assets/5663952/10369673/c14bf1da-6ddb-11e5-8471-1e2cad533c8a.png)
After 2 minutes: 9512.02 kByte/s

## New when copying files:
![new-with-copy](https://cloud.githubusercontent.com/assets/5663952/10369690/d5b47fac-6ddb-11e5-8404-c74453be6d3d.png)
After 2 minutes: 76382.85 kByte/s

However, it seems like this is also very Windows specific. It seems that Windows gives preference to file reads with the biggest buffer, so it really depends on which method you use for copying, and explains why small reads are penalized so much.

I tried remote -> remote, and the numbers are the same (612 vs 617KB/s).

# magic numbers

The thresholds are set to limit the downsides. They could of course be user specified, but I don't like pushing the "problem" on to the user. Mainly unneeded overhead on small files. Let's take them one by one:

## number of buffers
This is mainly a memory concern. More buffers are able to smooth out longer input drought or output bursts, but adding to many will eat too much memory. 3 or 4 seems reasonable.

## buffer size
Spinning disks likes sequential reads, so we want to make buffers a "reasonable" size. I choose 4MB, since that will allow the OS to read up to 4MB sequentially. Bigger buffers means more memory and longer latency - time from read request until data is ready for upstream. 1-16MB all seems reasonable.

Making buffers, say 32KB, will begin to give a significant overhead just by the number of channel send/receives needed to hand buffers between the goroutines. If we want that, we might as well slap in a bufio.Reader in and be done with it. In my experience `sync.Pool` doesn't help "enough" - it has significant synchronization overhead and the pool is flushed on every GC.

## threshold
Limit for when we don't expect to get much benefit. It is set rather high, but it is set where the 4MB initial read latency will have little overall effect. Could be anything from 1-50MB, but should reflect on the size of a buffer.

Either way, making anything less than a 10x adjustment on any of them will only have an extremely limited effect. The current magic numbers are meant to catch the "bad" effects and not display any significant downsides.
